External routing rules templates + regional presets support

runetfreedom opened the pull request:
This PR allows users to specify routing rule template sources. A template is just a version + `RoutingItem` array encoded in json (same logic as for builtin rules). Rules inside the template can be passed in two ways - through `url` field (then the rules will be downloaded automatically if necessary) and through `ruleSet` field. 

Since we have to download this template quite often (including at program startup), I preferred the shorter version with the url field in my template.

As for automatic downloading of geo files - now they are downloaded after changing the url in the configuration file and gui settings. For the method we discussed in #5832 (`guiNConfig.json` distributing) the user needs to download the geo files on the first run, otherwise routing rules will be downloaded that will fail because they use categories that are not in the default geo files. This will result in xray and sing-box not being able to run. It also seems logical that if the user has changed the source of the geo files, he wants to use them, so the program needs to download them.

However, there was a problem - in order for startup to realize that we now have geo files from a different source and need to download new ones, I need to know what source we actually have the files from right now, so I decided to save the current source along with the geo files to the text file `bin/geo.repo`.

I'm not sure if this is the best place to this, maybe you have ideas how this would be improved?

runetfreedom committed:
External routing rules templates + auto download geo file if repo changed

2dust commented:
Thanks PR.
Checking out this PR, there is too much interference with the original runtime process.
My suggestion is to add a separate function item in the settings menu, where the user clicks on it and downloads all the needed resources at once, including geo files and routing rules, etc;
This menu could be divided into countries, like Iran/Russia/...

- Menu Settings
- - download geo files and routing rules
- - -  Iran
- - -  Russia

runetfreedom commented:
Ok, then I will remove the code that automatically downloads geo files and add menu items, after that I will update the PR. 

Maybe we should call it “presets”?

runetfreedom committed:
Regional presets support

runetfreedom commented:
@2dust I updated PR, now instead of auto downloading it's just a button in the menu.

I also added a "Default" preset so that if the user clicked on a preset by mistake, they could easily go back.

2dust commented:
Some of the code will be adjusted to minimize the impact on the original process.
